src/material/stepper/testing/step-harness.ts

```diff
@@ -0,0 +1,116 @@
+  static hostSelector = '.mat-step-header';
+
+  /**
+   * Gets a `HarnessPredicate` that can be used to search for a `MatStepHarness` that meets
+   * certain criteria.
+   * @param options Options for filtering which steps are considered a match.
+   * @return a `HarnessPredicate` configured with the given options.
+   */
+  static with(options: StepHarnessFilters = {}): HarnessPredicate<MatStepHarness> {
+    return new HarnessPredicate(MatStepHarness, options)
+        .addOption('label', options.label,
+            (harness, label) => HarnessPredicate.stringMatches(harness.getLabel(), label))
+        .addOption('selected', options.selected,
+            async (harness, selected) => (await harness.isSelected()) === selected)
+        .addOption('completed', options.completed,
+            async (harness, completed) => (await harness.isCompleted()) === completed)
+        .addOption('invalid', options.invalid,
+            async (harness, invalid) => (await harness.hasErrors()) === invalid);
+  }
+
+  /** Gets the label of the step. */
+  async getLabel(): Promise<string> {
+    return (await this.locatorFor('.mat-step-text-label')()).text();
+  }
+
+  /** Gets the `aria-label` of the step. */
+  async getAriaLabel(): Promise<string|null> {
+    return (await this.host()).getAttribute('aria-label');
+  }
+
+  /** Gets the value of the `aria-labelledby` attribute. */
+  async getAriaLabelledby(): Promise<string|null> {
+    return (await this.host()).getAttribute('aria-labelledby');
+  }
+
+  /** Whether the step is selected. */
+  async isSelected(): Promise<boolean> {
+    const host = await this.host();
+    return (await host.getAttribute('aria-selected')) === 'true';
+  }
+
+  /** Whether the step has been filled out. */
+  async isCompleted(): Promise<boolean> {
+    const state = await this._getIconState();
+    return state === 'done' || (state === 'edit' && !(await this.isSelected()));
+  }
+
+  /**
+   * Whether the data inside the step has any errors. Note that this will only match if showing
+   * errors has been enabled through the `showErrors` option in `STEPPER_GLOBAL_OPTIONS`.
+   */
+  async hasErrors(): Promise<boolean> {
+    return (await this._getIconState()) === 'error';
+  }
+
+  /** Whether the step is optional. */
+  async isOptional(): Promise<boolean> {
+    // If the node with the optional text is present, it means that the step is optional.
+    const optionalNode = await this.locatorForOptional('.mat-step-optional')();
+    return !!optionalNode;
+  }
+
+  /**
+   * Selects the given step by clicking on the label. The step may not be selected
+   * if the stepper doesn't allow it (e.g. if there are validation errors).
+   */
+  async select(): Promise<void> {
+    await (await this.host()).click();
+  }
+
+  /**
+   * Gets a `HarnessLoader` that can be used to load harnesses for components within the step's
+   * content area.
+   */
+  async getHarnessLoaderForContent(): Promise<HarnessLoader> {
```

Comment:
can we use the new `ContentContainerComponentHarness` style for this?

Reply:
I was thinking about doing that, but all the querying in `ContentContainerComponentHarness` goes through `this.locatorFactory` which will only query under the host element, whereas the step harness needs to query through the `documentRootLocatorFactory`.

Reply:
Oh because the header and the content get separated, weird... in that case I would just override the methods and have them proxy through the same harness loader you're using here instead of the default one `ContentContainerComponentHarness` uses

Reply:
Done, although I don't feel great about creating the global CSS selector through string concatenation.